lib/read/lib/Cardano/Wallet/Read/Block.hs

```diff
@@ -1,18 +1,156 @@
 
 {-------------------------------------------------------------------------------
     Block type
 -------------------------------------------------------------------------------}
 -- | Type synonym for 'CardanoBlock' with cryptography as used on mainnet.
-type Block = O.CardanoBlock O.StandardCrypto
+type ConsensusBlock = O.CardanoBlock O.StandardCrypto
+
+-- Family of era-specific block types
+type family BlockT era where
+    BlockT ByronEra = O.ByronBlock
+    BlockT ShelleyEra =
+        O.ShelleyBlock (TPraos StandardCrypto) (O.ShelleyEra StandardCrypto)
+    BlockT AllegraEra =
+        O.ShelleyBlock (TPraos StandardCrypto) (O.AllegraEra StandardCrypto)
+    BlockT MaryEra =
+        O.ShelleyBlock (TPraos StandardCrypto) (O.MaryEra StandardCrypto)
+    BlockT AlonzoEra =
+        O.ShelleyBlock (TPraos StandardCrypto) (O.AlonzoEra StandardCrypto)
+    BlockT BabbageEra =
+        O.ShelleyBlock (Praos StandardCrypto) (O.BabbageEra StandardCrypto)
+    BlockT ConwayEra =
+        O.ShelleyBlock (Praos StandardCrypto) (O.ConwayEra StandardCrypto)
+
+newtype Block era = Block {unBlock :: BlockT era}
+
+-- | Get sequence of transactions in the block.
+txsFromBlockE :: EraFun Block ([] :.: Tx)
+txsFromBlockE =
+    EraFun
+        { byronFun = getTxs' getTxsFromBlockByron
+        , shelleyFun = getTxs' getTxsFromBlockShelleyAndOn
+        , maryFun = getTxs' getTxsFromBlockShelleyAndOn
+        , allegraFun = getTxs' getTxsFromBlockShelleyAndOn
+        , alonzoFun = getTxs' getTxsFromBlockShelleyAndOn
+        , babbageFun = getTxs' getTxsFromBlockShelleyAndOn
+        , conwayFun = getTxs' getTxsFromBlockShelleyAndOn
+        }
+  where
+    getTxs' f (Block block) = Comp $ Tx <$> f block
+
+getTxsFromBlockByron :: O.ByronBlock -> [TxT ByronEra]
+getTxsFromBlockByron block =
+    case Byron.byronBlockRaw block of
+        Byron.ABOBBlock b ->
+            map (() <$) . Byron.unTxPayload . Byron.blockTxPayload $ b
+        Byron.ABOBBoundary _ -> []
+
+getTxsFromBlockShelleyAndOn
+    :: (Shelley.EraSegWits era, EncCBOR (ShelleyProtocolHeader proto))
+    => O.ShelleyBlock proto era
+    -> [Ledger.Tx era]
+getTxsFromBlockShelleyAndOn (O.ShelleyBlock (Shelley.Block _ txs) _) =
+    toList (Shelley.fromTxSeq txs)
+
+-- | Convert block as received from cardano-node
+-- via Haskell library of mini-protocol.
+fromConsensusBlock :: ConsensusBlock -> EraValue Block
+fromConsensusBlock = \case
+    O.BlockByron b -> inject byron $ Block b
+    O.BlockShelley block -> inject shelley $ Block block
+    O.BlockAllegra block -> inject allegra $ Block block
+    O.BlockMary block -> inject mary $ Block block
+    O.BlockAlonzo block -> inject alonzo $ Block block
+    O.BlockBabbage block -> inject babbage $ Block block
+    O.BlockConway block -> inject conway $ Block block
+
+getTxs :: O.CardanoBlock StandardCrypto -> EraValue ([] :.: Tx)
+getTxs = applyEraFun txsFromBlockE . fromConsensusBlock
```

Comment:
```suggestion
getTxs :: O.CardanoBlock StandardCrypto -> [EraValue Tx]
getTxs = sequenceEraValue . applyEraFun txsFromBlockE . fromConsensusBlock
```

This definition seems more convenient in basic usage scenarios to me? 🤔 The basic scenario being "get all transactions in a block, don't care which eras they are in". (A more advanced scenario would work with the `EraFun`.)

Reply:
I removed that function in next PR, no use case.

Reply:
k